packages/dds/merge-tree/src/client.ts

```diff
@@ -346,6 +350,49 @@ export class Client {
         return this.mergeTree.getMarkerFromId(id);
     }
 
+    /**
+     * Revert an op
+     */
+    public rollback?(op: any, localOpMetadata: unknown) {
+        if (op.type !== MergeTreeDeltaType.INSERT) {
+            throw new Error("Unsupported op type for rollback");
+        }
+
+        const lastList = this.mergeTree.pendingSegments?.prev;
+        if (!lastList) {
+            throw new Error("No pending segments");
+        }
+        ListRemoveEntry(lastList);
+        const pendingSegmentGroup = lastList.data;
+        if (pendingSegmentGroup === undefined || pendingSegmentGroup !== localOpMetadata) {
+            throw new Error("Rollback op doesn't match last edit");
+        }
+        let length = 0;
+        for (const segment of pendingSegmentGroup.segments) {
+            const segmentSegmentGroup = segment.segmentGroups.dequeue();
+            assert(segmentSegmentGroup === pendingSegmentGroup, "Unexpected segmentGroup in segment");
+            if (TextSegment.is(segment)) {
+                length += segment.text.length;
+            } else if (Marker.is(segment)) {
+                length += 1;
+            } else {
+                throw new Error("Unsupported segment type for rollback");
+            }
+        }
+
+        const start = op.pos1 as number;
+        const segWindow = this.getCollabWindow();
+        const removeOp = createRemoveRangeOp(start, start + length);
+        this.mergeTree.markRangeRemoved(
```

Comment:
how does rollback work at the runtime level? Mainly I want to know: in the tests you've added, do the inserts inside `orderSequentially()` still get broadcast to other clients even though an error is thrown from within orderSequentially's callback?

Reply:
order sequentially will only send ops if it completed successfully. previously it would just close the container on error. with rollback we attempt to rollback each local operation, neither the original or the rollback will submit ops. it should be like it never happend

Reply:
OK, that's what I was guessing. I was worried this strategy seems like it might necessitate updating the model for seq bookkeeping and handling of segments at a given seq/refseq/localrefseq. I had the same concern as you on `TreeMaintenanceSequenceNumber`. Using `UniversalSequenceNumber` definitely seems better to me, but I think we might never end up zamboni'ing the segment due to this TODO in addToLRUSet:

```typescript
    private addToLRUSet(segment: ISegment, seq: number) {
        // If the parent node has not yet been marked for scour (i.e., needsScour is not false or undefined),
        // add the segment and mark the mark the node now.

        // TODO: 'seq' may be less than the current sequence number when inserting pre-ACKed
        //       segments from a snapshot.  We currently skip these for now.
        if (segment.parent!.needsScour !== true && seq > this.collabWindow.currentSeq) {
            segment.parent!.needsScour = true;
            this.segmentsToScour!.add({ segment, maxSeq: seq });
        }
    }
```

I feel like the sooner this only-locally-existing removed segment gets zambonied, the more confident I am that the eventual consistency model of merge tree isn't affected. So maybe worth investigating/testing the zamboni behavior that happens here.

Reply:
What should I check for verifying zamboni-ing? That any split segments should get merged and the removed segments go away? Like the rollback and edits within the orderSequentially never happened?

Reply:
Added a zamboni test